fiat-ldap-v2/src/main/java/com/netflix/spinnaker/fiat/roles/ldap/LdapUserRolesProviderV2.java

```diff
@@ -0,0 +1,202 @@
+    if (StringUtils.isEmpty(configProps.getGroupSearchBase())) {
+      return new ArrayList<>();
+    }
+
+    String fullUserDn = getUserFullDn(userId);
+
+    if (fullUserDn == null) {
+      // Likely a service account
+      log.debug("fullUserDn is null for {}", userId);
+      return new ArrayList<>();
+    }
+
+    String[] params = new String[] {fullUserDn, userId};
+    log.debug("Searching for groups using \ngroupSearchBase: {}\ngroupSearchFilter: {}\nparams: {}\ngroupRoleAttributes: {}",
+            configProps.getGroupSearchBase(),
+            configProps.getGroupSearchFilter(),
+            StringUtils.join(params, " :: "),
+            configProps.getGroupRoleAttributes());
+
+
+    // Copied from org.springframework.security.ldap.userdetails.DefaultLdapAuthoritiesPopulator.
+    Set<String> userRoles =
+        ldapTemplate.searchForSingleAttributeValues(
+            configProps.getGroupSearchBase(),
+            configProps.getGroupSearchFilter(),
+            params,
+            configProps.getGroupRoleAttributes());
+
+    log.debug("Got roles for user {}: {}", userId, userRoles);
+    return userRoles.stream()
+        .map(role -> new Role(role).setSource(Role.Source.LDAP))
+        .collect(Collectors.toList());
+  }
+
+  private class RoleFullDNtoUserRoleMapper implements AttributesMapper<Role> {
+    @Override
+    public Role mapFromAttributes(Attributes attrs) throws NamingException {
+      return new Role(attrs.get(configProps.getGroupRoleAttributes()).get().toString())
+          .setSource(Role.Source.LDAP);
+    }
+  }
+
+  private class UserRoleMapper implements AttributesMapper<Pair<String, Collection<Role>>> {
+    @Override
+    public Pair<String, Collection<Role>> mapFromAttributes(Attributes attrs)
+        throws NamingException {
+      List<Role> roles = new ArrayList<>();
+      Attribute memberOfAttribute = attrs.get("memberOf");
+      if (memberOfAttribute != null) {
+        for (NamingEnumeration<?> memberOf = memberOfAttribute.getAll(); memberOf.hasMore(); ) {
+          String roleDN = memberOf.next().toString();
+          LdapName ln = org.springframework.ldap.support.LdapUtils.newLdapName(roleDN);
+          String role =
+              org.springframework.ldap.support.LdapUtils.getStringValue(
+                  ln, configProps.getGroupRoleAttributes());
+          roles.add(new Role(role).setSource(Role.Source.LDAP));
+        }
+      }
+
+      return Pair.of(
+          attrs.get(configProps.getUserIdAttributes()).get().toString().toLowerCase(), roles);
+    }
+  }
+
+  @Override
+  public Map<String, Collection<Role>> multiLoadRoles(Collection<ExternalUser> users) {
+    if (StringUtils.isEmpty(configProps.getGroupSearchBase())) {
+      return new HashMap<>();
+    }
+    StringBuilder filter = new StringBuilder();
+    filter.append("(|");
+    users.forEach(
+        u -> filter.append(MessageFormat.format(configProps.getUserSearchFilter(), u.getId())));
+    filter.append(")");
```

Comment:
You can create a new `OrFilter` and run `filter.or(Filter)` on each filter to or together.